File: node/src/memory/durable.ts
import type { Memory, MemoryQuery, MemoryRecord, MemoryScope, MemorySearchOptions, MemoryStore } from "./protocols.js"

/**
 * Binds the host storage protocol to the public Memory contract for one agent and namespace.
 * This is deliberately an SDK adapter: it does not lower memory mutations into the Kernel.
 */
export class DurableMemory implements Memory {
  readonly namespace: string

  constructor(
    private readonly store: MemoryStore,
    private readonly agentId: string,
    private readonly scope: MemoryScope,
  ) {
    this.namespace = scope.namespace
  }

  async search(query: string, options: MemorySearchOptions = {}): Promise<MemoryRecord[]> {
    const request: MemoryQuery = {
      scope: this.scope,
      query,
      top_k: options.topK ?? 5,
      kinds: options.kinds ?? [],
      ...(options.minScore === undefined ? {} : { min_score: options.minScore }),
    }
    return (await this.store.search(this.agentId, request)).map(hit => hit.record)
  }

  async get(id: string): Promise<MemoryRecord | null> {
    const record = await this.store.get(this.agentId, id)
    return record !== null && sameScope(record.scope, this.scope) ? record : null
  }

  async put(record: MemoryRecord): Promise<void> {
    if (!sameScope(record.scope, this.scope)) {
      throw new Error("memory record scope must match the bound Memory scope")
    }
    await this.store.put(this.agentId, record)
  }

  async delete(id: string): Promise<void> {
    const record = await this.get(id)
    if (record !== null) await this.store.delete(this.agentId, id)
  }
}

function sameScope(left: MemoryScope, right: MemoryScope): boolean {
  return left.tenant_id === right.tenant_id && left.namespace === right.namespace
}

File: python/deepstrike/memory/durable.py
"""Public agent-bound durable memory adapter.

This contract is intentionally separate from ``WorkingMemory`` (an SDK scratch pad) and from
``MemoryStore`` (the runner's session/extraction protocol). It does not enter Kernel wire state.
"""
from __future__ import annotations

from deepstrike.memory.protocols import (
    MemoryQuery,
    MemoryRecord,
    MemoryScope,
    MemorySearchOptions,
    MemoryStore,
)


class DurableMemory:
    def __init__(self, store: MemoryStore, agent_id: str, scope: MemoryScope) -> None:
        self._store = store
        self._agent_id = agent_id
        self._scope = scope
        self.namespace: str | None = scope.namespace

    async def search(
        self,
        query: str,
        options: MemorySearchOptions | None = None,
        *,
        top_k: int | None = None,
    ) -> list[MemoryRecord]:
        resolved = options or MemorySearchOptions()
        request = MemoryQuery(
            scope=self._scope,
            query=query,
            top_k=top_k if top_k is not None else resolved.top_k,
            kinds=resolved.kinds,
            min_score=resolved.min_score,
        )
        return [
            hit.record
            for hit in await self._store.search(self._agent_id, request)
            if hit.record.scope == self._scope
        ]

    async def get(self, record_id: str) -> MemoryRecord | None:
        record = await self._store.get(self._agent_id, record_id)
        return record if record is not None and record.scope == self._scope else None

    async def put(self, record: MemoryRecord) -> None:
        if record.scope != self._scope:
            raise ValueError("memory record scope must match the bound Memory scope")
        await self._store.put(self._agent_id, record)

    async def delete(self, record_id: str) -> None:
        if await self.get(record_id) is not None:
            await self._store.delete(self._agent_id, record_id)
